Delta Manifests

jwakre opened the pull request:
This PR enables the consumption of delta manifests in the swupd client. Delta manifests are manifests that may be found inside delta packs and contain a list of files (from the current version) that have been changed in the TO version.

This PR also cherry picks some changes from #565 which are used by the added tests. Once these cherry picked changes are merged into master, they will be removed from this PR.

Closes #561

jwakre commented on src/update.c:
```diff
@@ -365,7 +365,45 @@ static int main_update()
+	}
+	set_subscription_versions(server_manifest, current_manifest, &latest_subs);
+
+	grabtime_start(&times, "Download Packs");
```
The "Recurse and Consolidate Manifests" time recording block starts before loading the server MoM, so now it includes pack downloading in its time statistics. I think that the "Recurse and Consolidate Manifests" time block should be moved so that it starts before loading the subscribed manifests instead of before loading the server MoM. I will open a separate issue for this.

jwakre replied:
Below is the timing issue:

#584

castulo commented on Makefile.am:
```diff
@@ -233,6 +233,10 @@ BATS = \
+	test/functional/update_v2/update-delta-manifest-invalid.bats \
+	test/functional/update_v2/update-delta-manifest-no-iterative.bats \
+	test/functional/update_v2/update-delta-manifest-and-full.bats \
+	test/functional/update_v2/update-delta-manifest.bats \
```
since there is already a PR that removes the update_v2 directory I'd suggest to add this directly to the update directory.

jwakre replied:
Good idea, I will update this.

castulo commented on test/functional/update_v2/update-delta-manifest.bats:
```diff
@@ -0,0 +1,58 @@
+	EOM
+	)
+	assert_in_output "$expected_output"
+}
```
What about the case when a bundle has updates during a couple of versions?
For example:

create_test_environment "$TEST_NAME" 10
create_bundle -L -n bundle1 -v 10 -f /file1 "$TEST_NAME"
create_version "$TEST_NAME" 20 10 staging
update_bundle -i "$TEST_NAME" bundle1 --update /file1
create_version "$TEST_NAME" 30 20 staging
update_bundle -i "$TEST_NAME" bundle1 --update /file1

So if the current version is 10 and we do an update, it should update all the way to version 30, iterative and delta manifests are going to exist for each version, but since the jump is larger than one version the delta manifests should not be used, correct? (I know the iterative manifest should not be used)
Maybe a test for this scenario is worth adding.

jwakre replied:
My understanding is that the server would look like the following:

* 20
    * pack-bundle1-from-0.tar
    * pack-bundle1-from-10.tar
* 30
    * pack-bundle1-from-0.tar
    * pack-bundle1-from-10.tar
    * pack-bundle1-from-20.tar

When updating FROM 10 TO 30, a valid delta manifest would be extracted from /30/pack-bundle1-from-10.tar. This delta manifest should include entries for the files that changed during the update FROM 10 TO 30.

jwakre replied:
@otaviobp Is my assumption wrong here? Are delta manifests only valid for the N-1 version?

If delta manifests only work for the N-1 verison, I will need to not look for old delta manifests.

matthewrsj replied:
Yes @jwakre right now we are only publishing delta manifests for bundle->previous (not necessarily N-1 or N-10)

matthewrsj replied:
Wait, no that's wrong. Delta manifests are valid for the FROM version to the TO version. I was getting my iterative and delta manifests mixed up again (I've been doing that a lot). If you download a pack from FROM (to your TO) it will have all the records of files used in deltas.

matthewrsj replied:
tl;dr @jwakre is right

jwakre replied:
Ok, thanks for the clarification. I think a test case for the scenario when a bundle has been updated more than once since an update, so I will add that.

jwakre commented on src/manifest.c:
```diff
@@ -679,38 +684,54 @@ struct manifest *load_mom_err(int version, bool latest, bool mix_exists, int *er
-	}
+	if (delta_manifest) {
+		if (file->peer != NULL) {
+			delta_manifest_version = file->peer->last_change;
```
I am going to condense this to one condition statement:

if (delta_manifest && file->peer != NULL), use delta manifest
else, don't use delta manifest

I can also check for the case when the manifest is unchanged and skip attempting to load a delta manifest.

jwakre commented on src/manifest.c:
```diff
@@ -1010,7 +1031,7 @@ struct list *recurse_manifest(struct manifest *manifest, struct list *subs, cons
 		}
 
-		sub = load_manifest(version1, version2, file, manifest, false);
+		sub = load_manifest(version1, version2, file, manifest, false, delta_manifest);
```
This attempts to load bundles that have not changed. The load_manifest function will fail to find an unchanged delta manifest and fallback to downloading the entire bundle manifest. I will add a case to skip bundles that have not changed when looking for delta_manifests.

jwakre commented:
I added a change so that when loading FROM bundles, unchanged bundles will be skipped. Earlier, unchanged bundles would attempt to download a delta manifest and fallback to download a full manifest.

jwakre commented:
Based on offline feedback from @matthewrsj to replace functions that take a boolean delta_manifest argument with wrapper functions, I intend to do the following:

1. Instead of passing the delta_manifest argument to the recurse_manifest and load_manifest functions, I will add a use_delta_manifest variable to the manifest struct which will only be set to true for the FROM manifest.

2. I will create a wrapper function for the manifest_from_file function to abstract loading a delta manifest vs loading a normal manifest.

jwakre commented:
> Based on offline feedback from @matthewrsj to replace functions that take a boolean delta_manifest argument with wrapper functions, I intend to do the following:
> 
> 1. Instead of passing the delta_manifest argument to the recurse_manifest and load_manifest functions, I will add a use_delta_manifest variable to the manifest struct which will only be set to true for the FROM manifest.

The change described above has been pushed. Adding the 'use_delta_manifest' variable removes most of the API changes.

matthewrsj commented on src/clean.c:
```diff
@@ -203,6 +203,16 @@ static bool is_manifest_delta(const char UNUSED_PARAM *dir, const struct dirent
+{
+	static const char delta_str[] = ".D.";
+
+	if (is_manifest(dir, entry) && strstr(entry->d_name, delta_str) != NULL) {
```
this is a `bool` return function. Just do
```c
return is_manifest(dir, entry) && strstr(entry->d_name, delta_str) != NULL;
```

matthewrsj commented on src/clean.c:
```diff
@@ -288,14 +299,22 @@ static int clean_staged_manifests(const char *path, bool dry_run, bool all)
 
+		/* Delta manifests are only valid when updating FROM the current version
+		 * TO the new version and should not be cached. */
+		string_or_die(&version_dir, "%s/%s", state_dir, name);
```
`version_dir` gets reassigned to the same thing again before `remove_if(version_dir, dry_run, is_manifest)` below. Remove the reassignment down there.

matthewrsj replied:
This also changes a bit in #574 which should be merged soon, so this will need an update.

matthewrsj commented on src/manifest.c:
```diff
@@ -110,7 +110,12 @@ static struct manifest *manifest_from_file(int version, char *component, bool he
-	string_or_die(&filename, "%s/%i/Manifest.%s", basedir, version, component);
+
+	if (delta_manifest_version > 0) {
+		string_or_die(&filename, "%s/%i/Manifest.%s.D.%i", basedir, delta_manifest_version, component, version);
```
Leave a comment here that delta manifests are extracted to the `delta_manifest_version` version directory (which is the `last_change` version) but have the `version` in the name.

matthewrsj commented on src/manifest.c:
```diff
@@ -683,34 +689,48 @@ struct manifest *load_manifest(int current, int version, struct file *file, stru
+		if (ret != 0) {
+			if (retried == false) {
+				remove_manifest_files(file->filename, version, file->hash);
+				retried = true;
```
/me really wants #211 to happen...

matthewrsj commented on src/manifest.c:
```diff
@@ -683,34 +689,48 @@ struct manifest *load_manifest(int current, int version, struct file *file, stru
+			} else {
+				remove_manifest_files(file->filename, version, file->hash);
+				retried = true;
+			}
```
ok, so if manifest failed to load and retried is false, we set delta_manifest to false, delta_manifest_version to 0 (essentially turning off delta_manifest options for this manifest) and retry without setting retried to true. This means we will try to download the full manifest and still allow it to retry one more time.

This flow is good, but please add a lengthy comment explaining the flow. It took me a while to figure that out just by reading the code here.

matthewrsj commented on src/manifest.c:
```diff
@@ -801,6 +823,17 @@ struct list *create_update_list(struct manifest *server)
+		 * so files in the TO file list without a peer are not new when the file
+		 * exists on the system. */
+		filename = mk_full_filename(path_prefix, file->filename);
+		ret = verify_file_lazy(filename);
```
we talked about just using the TO file list as "new or changed" files and combining that metric into one report. Is that feasible to do?

jwakre replied:
We can already count changed files because delta manifests will have entries for them. The problem here is that unchanged files will not be tracked in the FROM file list when using delta manifests. As a result, the old method could not differentiate unchanged files from new ones.

I don't think combining unchanged and new files would be useful.

matthewrsj commented on src/manifest.c:
```diff
@@ -866,13 +900,30 @@ void link_manifests(struct manifest *m1, struct manifest *m2)
+		 * so files in the TO file list without a peer are only new when they
+		 * don't exist on the system. */
+		filename = mk_full_filename(path_prefix, file2->filename);
+		if (!verify_file_lazy(filename)) {
```
Same question as above.

matthewrsj commented on src/manifest.c:
```diff
@@ -866,13 +900,30 @@ void link_manifests(struct manifest *m1, struct manifest *m2)
+		/* Delta manifests don't include unchanged files in the FROM file list,
+		 * so files in the TO file list without a peer are only new when they
+		 * don't exist on the system. */
+		filename = mk_full_filename(path_prefix, file2->filename);
```
Same question as above, but anyways at this point you are repeating code a lot. Make this its own function if you end up keeping it.

otaviobp commented on src/manifest.c:
```diff
@@ -801,6 +803,17 @@ struct list *create_update_list(struct manifest *server)
+		/* Delta manifests don't include unchanged files in the FROM file list,
+		 * so files in the TO file list without a peer are not new when the file
+		 * exists on the system. */
+		filename = mk_full_filename(path_prefix, file->filename);
```
it looks like this comment doesn't make sense anymore after Matthews comment

otaviobp replied:
My suggestion isn't going to work. If an old bundle is included in a new version of a new bundle, the version of the file will be older than current version and it will be installed anyway.

I was trying to understand the best approach to that and I wrote a patch, but I still have 2 tests failing. https://github.com/otaviobp/swupd-client/commit/fcbf00934f13fde580470cbaad7ce7e23a93c380

jwakre replied:
I think we can solve this case by doing the following:

1. Add a variable to the file struct that is set to true when a file is included

2. Initialize variable for loaded files to false

3. Mark files pulled in by included bundles true (in add_subscriptions).

4. Check for included files with an older version than the current.

jwakre replied:
Since we can't use delta manifests to decide information about file updates, I no longer see a need for the link_manifests function. I think we could consolidate the file statistic counting to create_update_list and remove the link_manifests function.

matthewrsj commented:
> I have a feeling that a great refactor in create_update_list() is needed to avoid downloading the current manifest from server. We could use current version of the system to decide if we need to update a file or not instead of using the version of that file in the current manifest.

Very good point @otaviobp. That block of code was put in to optimize rename detection iirc, but we don't do things the same way any more. Regardless, if we refactor that code it will need careful review.

matthewrsj commented:
Going to comment here about a few cases that need to be handled when not checking against the previous manifest and instead using the version of the installed OS.
1. file in the to manifest is a newer version than the OS, **add to update list**.
2. file in the to manifest is an older version than the OS, but the manifest was added as a new include and was not tracked on the system, and the file itself is only provided by the newly included bundle, if the file is *missing* or *invalid* on the file system, **add to update list**.
3. file in the to manifest is an older version than the OS and the manifest was already tracked on the file system, **do not add to update list** regardless of if that file is *also* provided by a newly included bundle.

matthewrsj commented:
A note on minversion bumps:
```
13:59 < mrsj> the only time a file will get an updated version number with no hash or content change or file name change is when you have a minversion update
14:00 < mrsj> which raises an issue...
14:00 < mrsj> our current plan will cause ALL OS files to be downloaded for a minversion update
14:00 < mrsj> because a minversion marks all files with a new version
14:00 < mrsj> :/
14:01 < mrsj> so our plan would put those under Case1, where the file is greater than the current OS version
14:01 < mrsj> so for every tracked file you will be adding those to the manifest
14:01 < mrsj> CURRENTLY...
14:01 < mrsj> in manifest.c,
14:01 < mrsj> we can detect minversion updates because we know the file version changed by the hash and flags stayed the same
14:01 < mrsj> can't do that when we aren't using the old manifests any more
14:02 < mrsj> we need another solution
14:02 < mrsj> one solution:
14:02 < mrsj> add a manifest header field to the MoM
14:02 < mrsj> minversion:\t<minversion>
14:02 < mrsj> and if <minversion> is greater than current OS version we do the fallback to verifying all files with versions greater than our own OS version
14:03 < mrsj> make sense jwakre, otaviobp?
14:03 < mrsj> this is unfortunate
14:03 < mrsj> but we should be able to add a minversion field to the MoM in mixer without requiring a format bump for the client
14:03 < mrsj> because swupd ignores header fields it doesn't know about
14:04 < jwakre> What is a minversion?
14:05 < mrsj> a minversion is when all content is "updated" on the server-side
14:05 < mrsj> by mixer
14:05 < otaviobp> mrsj: this sounds correct to me
14:05 < mrsj> so all file versions are updated to that "minversion" build regardless of whether they actually physically changed or not
14:06 < jwakre> oh, so a minimum version for all files on the system
14:06 < mrsj> right now in manifest.c (search for "minversion") the client falls back to verifying all files in the update against the system
14:06 < otaviobp> gtkramer said they usually do it once a mounth, so it's releavant to be optimized for that
14:06 < mrsj> to prevent having to download the ENTIRE OS
14:06 < otaviobp> I thought that this was a rare thing, but it's not
14:06 < mrsj> no it is often
14:06 < mrsj> because it allows devops to free up space on their builder
14:06 < mrsj> :/
14:06 < mrsj> apparently disk space is not cheap
14:06 < otaviobp> disk isn't that cheap, apparently 
14:07 < mrsj> despite what everyone tells us ;P
14:09 < mrsj> that was a lot of writing... I should have put it in the PR...
14:09  * mrsj copies
```

jwakre commented:
PR Updates:

1. Rebased on to master_new_manifests branch which includes iterative manifests and the refactored update list.

2. @otaviobp's feedback was addressed by the refactored update list PR which has already been committed.

3. @matthewrsj 's feedback:
* Added comments for delta manifest flow
* Removed unnecessary version_dir assignment in clean.c
* Removed condition statement in is_delta_manifest function in clean.c
* "TO" file list feedback was addressed by refactored update list PR.

castulo commented on src/update.c:
```diff
@@ -447,38 +469,20 @@ static int main_update()
 	server_manifest->files = files_from_bundles(server_manifest->submanifests);
 	server_manifest->files = consolidate_files(server_manifest->files);
 
-	set_subscription_versions(server_manifest, current_manifest, &update_subs);
```
cool

castulo commented on test/functional/testlib.bash:
```diff
@@ -2071,6 +2071,10 @@ update_bundle() {
+
+		# TODO: Delta packs should be generated for an update from the current version
+		#       to the latest version. If a bundle is updated more than once in a single
+		#       update, an incorrect delta pack will be generated. Issue 615
```
good catch

jwakre commented on src/update.c:
```diff
@@ -430,6 +436,7 @@ static int main_update()
 	 * First load up the old (current) manifests. Statedir could have been cleared
 	 * or corrupt, so don't assume things are already there. Updating subscribed
 	 * manifests is done as part of recurse_manifest */
+	link_submanifests(current_manifest, server_manifest, initial_subs, initial_subs, false);
```
To find where a delta manifest is extracted to inside the state dir, I need to know its corresponding peer bundle's last change. The link_submanifests function must be called earlier so that the peer is set before I look for the delta manifest.

The peer->last_change is used to set the 'delta_manifest_state_dir' variable in load_manifests.

jwakre replied:
I could move this above the comment though...

castulo commented on test/functional/update/update-delta-manifest-and-full.bats:
```diff
@@ -0,0 +1,60 @@
+		Update successful. System updated from version 10 to version 20
+	EOM
+	)
+	assert_in_output "$expected_output"
```
if possible is better to use _assert_is_output_ than _assert_in_output_

castulo commented on test/functional/update/update-delta-manifest-invalid.bats:
```diff
@@ -0,0 +1,51 @@
+	update_bundle -i "$TEST_NAME" bundle1 --update /file1
+
+	# Overwrite delta manifest with invalid file
+	sudo sh -c "echo 1 > $WEBDIR/20/20/Manifest.bundle1.D.10"
```
it is ok like tit is now, but just FYI there is a function in the test library to do that:
```
Usage:
    write_to_protected_file [-a] <file> <stream>

Options:
    -a    If used the text will be appended to the file, otherwise it will be overwritten
```

castulo commented on test/functional/update/update-delta-manifest-invalid.bats:
```diff
@@ -0,0 +1,51 @@
+		Update successful. System updated from version 10 to version 20
+	EOM
+	)
+	assert_in_output "$expected_output"
```
same comment as above

jwakre commented:
Thanks for the feedback @castulo. I made the following changes to address your feedback:

* Moved link_submanifests above the recurse_manifest comment
* Changed assert_in_output -> assert_is_output for added tests
* Replaced the echo file write with write_to_protected_file

otaviobp commented:
Just wait for @matthewrsj to check if he has something else to add and then we can merge this.
